Add sample_weight to the calculation of alphas in enet_path and LinearModelCV

s-banach opened the pull request:
#### Reference Issues/PRs
Fixes #22914.

#### What does this implement/fix? Explain your changes.
Modifies `_alpha_grid` function in `linear_model._coordinate_descent` to accept a `sample_weight` argument.

The function `_alpha_grid` is called in two places, `enet_path` and `LinearModelCV`.
The new `sample_weight` argument is not used by `enet_path`, but it is used by `LinearModelCV`.

#### Any other comments?
Since my previous PR on this issue, `_preprocess_data` has been rewritten.

s-banach committed:
Update _alpha_grid to take sample_weight

It seems like this single call to _preprocess_data suffices in all cases.

s-banach committed:
Add a simple test for alpha_max with sample_weight

This tiny example was given in https://github.com/scikit-learn/scikit-learn/issues/22914.
The test merely asserts that alpha_max is large enough to force the coefficient to 0.

lorentzenchr commented on sklearn/linear_model/_coordinate_descent.py:
```diff
@@ -146,47 +149,35 @@ def _alpha_grid(
             "your estimator with the appropriate `alphas=` "
             "argument."
         )
-    n_samples = len(y)
```
Could you keep `n_samples` and use it instead of `X.shape[0]`? Maybe even
```python
n_samples = X.shape[0]
```

s-banach replied:
Done.

lorentzenchr commented on sklearn/linear_model/_coordinate_descent.py:
```diff
@@ -146,47 +149,35 @@ def _alpha_grid(
-
-    sparse_center = False
-    if Xy is None:
-        X_sparse = sparse.isspmatrix(X)
-        sparse_center = X_sparse and (fit_intercept or normalize)
-        X = check_array(
-            X, accept_sparse="csc", copy=(copy_X and fit_intercept and not X_sparse)
```
I think this is valuable input checking that we should keep.

s-banach replied:
(1) `_alpha_grid` is only called from `LinearModelCV` (which always checks `X` and `y`) or `enet_path`, which also checks them if `check_input is True`.
(2) For this reason, I'm calling `_preprocess_data` with `check_input=False`. Even if you wanted to repeat the checking, it seems like we would just let `_preprocess_data` handle the checking.

lorentzenchr replied:
ok

lorentzenchr commented on sklearn/linear_model/_coordinate_descent.py:
```diff
@@ -146,47 +149,35 @@ def _alpha_grid(
-            )
-            mean_dot = X_offset * np.sum(y)
+        if sample_weight is not None:
+            yw = y * sample_weight / sample_weight.mean()
```
```suggestion
            yw = y * sample_weight
```
Otherwise, the unweighted case should read `yw = y / n_samples` instead of `yw = y` because it must be equivalent to all sample weights equal `1`.

s-banach replied:
I chose `sample_weight.mean()` rather than `sample_weight.sum()` for this reason.

s-banach replied:
I'll be honest with you. I'm just dividing by `n_samples` because it's what the old code did, and because it passes the tests.
There are some comments in the `ElasticNet` class that seem to explain it:
```python
# But enet_path coordinate descent minimizes
#
#     1/2 * sum(squared error) + alpha' * penalty                  (2)
#
# and therefore sets
#
#     alpha' = n_samples * alpha                                   (3)
```

lorentzenchr commented on sklearn/linear_model/_coordinate_descent.py:
```diff
@@ -146,47 +149,35 @@ def _alpha_grid(
+        else:
+            yw = y
+        if sparse.issparse(X):
+            Xy = safe_sparse_dot(X.T, yw, dense_output=True) - np.sum(yw) * X_offset
```
What about `X_scale` in case of `normalize = True`?

lorentzenchr commented on sklearn/linear_model/tests/test_coordinate_descent.py:
```diff
@@ -1642,6 +1642,18 @@ def test_enet_cv_grid_search(sample_weight):
     assert reg.alpha_ == pytest.approx(gs.best_params_["alpha"])
 
 
+@pytest.mark.parametrize("sparseX", [False, True])
```
I would also parametrize `fit_intercept`

lorentzenchr commented on sklearn/linear_model/tests/test_coordinate_descent.py:
```diff
@@ -1642,6 +1642,18 @@ def test_enet_cv_grid_search(sample_weight):
+    sample_weight = np.array([10, 1, 10, 1])
+    if sparseX:
+        X = sparse.csc_matrix(X)
+    reg = ElasticNetCV(n_alphas=1, cv=2).fit(X, y, sample_weight=sample_weight)
```
```suggestion
    reg = ElasticNetCV(n_alphas=1, eps=1, cv=2).fit(X, y, sample_weight=sample_weight)
```

lorentzenchr reviewed (commented):
A first round of review comments.

s-banach committed:
Update test

As per reviewer's suggestions:
(1) Clarify eps=1.
(2) Parameterize `fit_intercept`.

s-banach committed:
Clarify _alpha_grid.

(1) Give the name `n_samples` to the quantity `X.shape[0]`.
(2) Clarify that `y_offset` and `X_scale` are not used, since these are already applied to the data by `_preprocess_data`.

s-banach committed:
Clarify notation

lorentzenchr commented:
@TomDLT May I kindly ping you as your help would be much appreciated.

TomDLT commented on sklearn/linear_model/_coordinate_descent.py:
```diff
@@ -146,47 +149,35 @@ def _alpha_grid(
-        sparse_center = X_sparse and (fit_intercept or normalize)
-        X = check_array(
-            X, accept_sparse="csc", copy=(copy_X and fit_intercept and not X_sparse)
+    if Xy is not None and ((sample_weight is None) or (fit_intercept is False)):
```
Maybe we should warn if `Xy` is not None but is never used.
WDYT?

s-banach replied:
Fixed so that Xy is always used when it is not None.

s-banach replied:
My old code was probably correct. My memory just isn't so good, so I forgot why exactly I coded it that way.
I'm not even sure if it's even possible for `Xy` *not* to be `None`.
It's honestly kind of hard to read the linear_model source code and figure out what's going on.

TomDLT commented on sklearn/linear_model/_coordinate_descent.py:
```diff
@@ -146,47 +149,35 @@ def _alpha_grid(
-    alpha_max = np.sqrt(np.sum(Xy**2, axis=1)).max() / (n_samples * l1_ratio)
+    if Xyw.ndim == 1:
+        Xyw = Xyw[:, np.newaxis]
+    n_samples = X.shape[0]
```
Are you sure it is `n_samples` here, and not `sample_weight.sum()`?

s-banach replied:
Above in the code, we define `yw = y * sample_weight / sample_weight.mean()`. This allows us to define `n_samples` in a way that works regardless of whether sample weights are used.

(More generally, I'm willing to bet sklearn will eventually make the package-wide choice to normalize all sample weights to have mean 1. There seems to be some foot dragging on this matter for the sake of "backward-compatibility", despite the fact that almost every major release breaks backward-compatibility in some way or another....)

s-banach replied:
Removed the previous `sample_weight / sample_weight.mean()` so that we can use `sample_weight.sum()` here.

TomDLT reviewed (commented):
Looks good, although I did not check the math.

**Main remark:** The new test function tests that the computed `alpha_max` is larger or equal to the true `alpha_max`. To test that they are actually equal, we could test that `alpha_max * 0.99` does not return all-zero coefficients.

We could also add a test that the computation still works without sample weights.

s-banach committed:
Use Xy if it is provided.

s-banach commented:
> **Main remark:** The new test function tests that the computed `alpha_max` is larger or equal to the true `alpha_max`. To test that they are actually equal, we could test that `alpha_max * 0.99` does not return all-zero coefficients.

Does sklearn testing have a function like `assert_not_allclose`?

s-banach committed:
Update test, check alpha_max is not too large

s-banach commented:
> **Main remark:** The new test function tests that the computed `alpha_max` is larger or equal to the true `alpha_max`. To test that they are actually equal, we could test that `alpha_max * 0.99` does not return all-zero coefficients.

I have attempted to update the test according to your recommendation. Please let me know if it looks correct.

> We could also add a test that the computation still works without sample weights.
My feeling is that `test_enet_cv_sample_weight_consistency` basically guarantees this already.
Let me know your thoughts.

s-banach committed:
Fix test that alpha_max is not too large.

s-banach commented:
The main thing I'm confused about, is why it's even possible for `_alpha_grid` to be called before `X` and `y` are appropriately scaled by `sample_weight`. It seems that `enet_path` or `LinearModelCV` or something should be refactored such that the call to `_preprocess_data` can be removed from `_alpha_grid`.

TomDLT commented:
> My feeling is that `test_enet_cv_sample_weight_consistency` basically guarantees this already.

I don't think it guarantees that the `alpha_max` computation is correct. To do so, we could add a `@pytest.mark.parametrize("sample_weight", [[10, 1, 10, 1], None])` to the new test.

> The main thing I'm confused about, is why it's even possible for `_alpha_grid` to be called before `X` and `y` are appropriately scaled by `sample_weight`.

It seems weird indeed. It seems that `_pre_fit` is called either before `_alpha_grid` in `enet_path`, or after `_alpha_grid` in `LinearModelCV.fit` (in `_path_residuals`). We should clarify the situation.

s-banach committed:
Test alpha_max without sample_weight.

s-banach commented:
Per your suggestion, I parameterized the new test by `sample_weight`.

I know my opinion on this matter isn't very valuable, but I'll share anyway.
As you say, `_alpha_grid` can currently be called from two contexts: from a `path` method such as `enet_path`, or from a `LinearModelCV`.
Currently, `LinearModelCV` works by finding the best `alpha` using CV, then feeding that `alpha` into the non-CV version of the estimator. Instead, I think `LinearModelCV` should *begin* by using `path` to fit the full dataset. This will allow the user to see the full `coef_path` after fitting the model, and it may even improve the total runtime due to warm starting.
If this change is made, then `_alpha_grid` will only ever be called from within a `path` method. Thus `_alpha_grid` will not be asked to deal with `sample_weight` at all.

TomDLT commented:
I agree it makes more sense to compute the alpha grid within the path function. We would need to have `_path_residuals` return the computed alphas, but this is not a problem because the function is private.

We might still need to have `_alpha_grid` deal with `sample_weights` in the sparse case.

jeremiedbb commented:
We won't have time to review this one before the 1.2 release. Moving it to 1.3

jeremiedbb commented:
(didn't mean to close :/ )
